File: TP3/PilhaFlex.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>

typedef struct Jogador{
    int id;
    char nome[50];
    int altura;
    int peso;
    char universidade[80];
    int anoNascimento;
    char cidadeNascimento[50];
    char estadoNascimento[50];
}Jogador;
//metodo para dividir a string linha em substrings entre as virgulas
void split(char linha[], char substrings[8][100]) {
    int qtSubstrings = 0;
    int cS = 0;//posicao da substring atual
    int c = 0;//posicao da linha
    //inicializacao da matriz substrings para fins de controle
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 100; j++) {
            substrings[i][j] = '\0';
        }
    }
    //loop que repete ate que a string linha seja totalmente percorrida
    while (linha[c] != '\0') {
        if (linha[c] != ',') {
            while (linha[c] != ',' && linha[c] != '\0') {
                if(linha[c] == '\n')c++;//ignorar as quebras de linha
                else{
                    substrings[qtSubstrings][cS] = linha[c];
                    c++;
                    cS++;
                }
            }
            cS = 0;
            qtSubstrings++;
        } else {
            //condicional para caso o campo esteja vazio
            if(linha[c+1] == ','||linha[c+1] == '\n'||linha[c+1] == '\0'){
                strcpy(substrings[qtSubstrings],"nao informado");
                qtSubstrings++;
            }
            c++; 
        }
    }
}
//metodo para realizar a leitura de um arquivo e guardar as informacoes em um array de jogadores
void ler(Jogador jogadores[]){
    FILE *file=fopen("/tmp/players.csv","r");
    char linha[200];
    int qtJogadores=-1;//inicializacao negativa para que a primeira linha seja ignorada

    while(fgets(linha, sizeof(linha), file) != NULL){
        char substrings[8][100];
        if(qtJogadores>=0){
            split(linha,substrings);
            //conversao de strings para inteiros
            int ID = atoi(substrings[0]);
            int h = atoi(substrings[2]);
            int p = atoi(substrings[3]);
            int ano = atoi(substrings[5]);

            jogadores[qtJogadores].id=ID;
            strcpy(jogadores[qtJogadores].nome, substrings[1]);
            jogadores[qtJogadores].altura=h;
            jogadores[qtJogadores].peso=p;
            strcpy(jogadores[qtJogadores].universidade, substrings[4]);
            jogadores[qtJogadores].anoNascimento=ano;
            strcpy(jogadores[qtJogadores].cidadeNascimento, substrings[6]);
            strcpy(jogadores[qtJogadores].estadoNascimento, substrings[7]);
            qtJogadores++;
        
        }
        else qtJogadores++;
    }
    fclose(file);
}
typedef struct Celula{
   struct Celula *prox;
    Jogador elemento;
}Celula;

typedef struct PilhaFlex{
    Celula *topo;
}PilhaFlex;

//Metodo para preencher uma nova celula
void novaCelula(Celula *celula, Jogador jogador){
    celula->prox=NULL;
    celula->elemento=jogador;
    
}
//Metodo para inicializar a pilha
void inicializar(PilhaFlex *pilha){
    pilha->topo=NULL;
}

//Metodo para inserir no fim
void inserirFim(PilhaFlex *pilha,Jogador lido){
    Celula *celula=(Celula*)malloc(sizeof(Celula));
    novaCelula(celula,lido);
    Celula* aux=pilha->topo;
    pilha->topo=celula;
    pilha->topo->prox=aux;
    aux=NULL;
}
//Metodo para remover no fim
Jogador removerFim(PilhaFlex *pilha){
    Jogador jogador=pilha->topo->elemento;
    Celula *aux=pilha->topo;
    pilha->topo=pilha->topo->prox;
    aux=NULL;
    return jogador;
}

//metodo calcular o numero de registros da pilha
int tamanho(PilhaFlex *pilha){
    Celula *i=pilha->topo;
    int c=0;
    while(i!=NULL){
        c++;
        i=i->prox;
    }
    return c;
}

//metodo para imprimir os atributos dos jogadores presentes na pilha 
void imprime(PilhaFlex *pilha,Celula *i,int j){
    if(i!=NULL){
        imprime(pilha,i->prox,j-1);
        printf("[%d] ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n",
        j,
        i->elemento.nome,
        i->elemento.altura,
        i->elemento.peso,
        i->elemento.anoNascimento,
        i->elemento.universidade,
        i->elemento.cidadeNascimento,
        i->elemento.estadoNascimento);
        i=i->prox;
    }
}

int main(){
    PilhaFlex pilha;
    inicializar(&pilha);
    Jogador jogadores[3922];
    char id[10];
    char comandos[100];
    int nOperacoes=0;
    Jogador jogadoresR[100];
    int qtR=0;
    //construçao do subarray com os ids digitados
    do {
        scanf("%s", id);
        if (strcmp(id, "FIM") != 0) {
            int identificador = atoi(id);
            ler(jogadores);
            Jogador aux=jogadores[identificador];
            inserirFim(&pilha,aux);
        }
    } while (strcmp(id, "FIM") != 0);
    scanf("%d", &nOperacoes);
    //loop para receber todas as operaçoes
    for(int i=0;i<nOperacoes;i++){
        fflush(stdin);
        scanf(" %[^\r\n]%*c", comandos);
        //Repartiçao da string pelo delimitador espaço em branco
        char *tokens[100]; 
        int token_count = 0;
        char *token = strtok(comandos, " ");
        while (token != NULL) {
            tokens[token_count] = token;
            token_count++;
            token = strtok(NULL, " ");
        }
        
        if(strcmp(tokens[0],"R")==0){
            jogadoresR[qtR]=removerFim(&pilha);
            qtR++;
        }
        else if(strcmp(tokens[0],"I")==0){
            int identificador = atoi(tokens[1]);
            Jogador aux=jogadores[identificador];
            inserirFim(&pilha,aux);
            
        }
    }
    //impressao dos removidos
    for(int i=0;i<qtR;i++){
        printf("(R) %s \n",jogadoresR[i].nome);
    }
    int tam=tamanho(&pilha);
    imprime(&pilha,pilha.topo,tam-1);

}

File: Atividades/TP3/PilhaFlex.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Scanner;

class Jogador {
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;

    // Construtor vazio
    public Jogador() {
    }

    // Construtor com parâmetros
    public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento) {
        this.id = id;
        this.nome = nome;
        this.altura = altura;
        this.peso = peso;
        this.universidade = universidade;
        this.anoNascimento = anoNascimento;
        this.cidadeNascimento = cidadeNascimento;
        this.estadoNascimento = estadoNascimento;
    }

    // Métodos getters e setters
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public int getAltura() {
        return altura;
    }

    public void setAltura(int altura) {
        this.altura = altura;
    }

    public int getPeso() {
        return peso;
    }

    public void setPeso(int peso) {
        this.peso = peso;
    }

    public String getUniversidade() {
        return universidade;
    }

    public void setUniversidade(String universidade) {
        if(universidade==null){
            universidade="nao informado";
        }
        this.universidade = universidade;
    }

    public int getAnoNascimento() {
        return anoNascimento;
    }

    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }

    public String getCidadeNascimento() {
        return cidadeNascimento;
    }

    public void setCidadeNascimento(String cidadeNascimento) {
        if(cidadeNascimento==null){
            cidadeNascimento="nao informado";
        }
        this.cidadeNascimento = cidadeNascimento;
    }

    public String getEstadoNascimento() {
        return estadoNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento) {
        if(estadoNascimento==null)estadoNascimento="nao informado";
        this.estadoNascimento = estadoNascimento;
    }

    // Método clone
    public Jogador clone() {
        Jogador resp=new Jogador();
        resp.id=this.id;
        resp.nome=this.nome;
        resp.altura=this.altura;
        resp.peso=this.peso;
        resp.universidade=this.universidade;
        resp.anoNascimento=this.anoNascimento;
        resp.cidadeNascimento=this.cidadeNascimento;
        resp.estadoNascimento=this.estadoNascimento;
        return resp;
    }

    // Método ler que funciona através da extração de dados pelo arquivo csv
    public static Jogador[] ler() {
        Jogador jogadores[]=new Jogador[4000];
        int qtJ=-1; 
        String nomeArquivo = "/tmp/players.csv";
        //inicializado com -1 para que o loop while seja executado ignorando a primeira linh
        try (BufferedReader leitor = new BufferedReader(new FileReader(nomeArquivo))) {
            String linha;//guarda o conteudo da linha lida 
            while ((linha = leitor.readLine()) != null) {
                
                if(qtJ>=0){//Apartir da 2 linha
                    
                    jogadores[qtJ]=new Jogador();
                    Scanner sc= new Scanner(linha);
                    sc.useDelimiter(",");
                    int c=0;//contador de atributos
                    String atributos[]=new String[8];
                    while (sc.hasNext()) {
                        String str = sc.next();
                        atributos[c]=str;// armazena as substrings
                        c++;
                    }
                    if(atributos[6].length()==0){ 
                        atributos[6]=null;
                    }
                    if(atributos[4].length()==0){ 
                        atributos[4]=null;
                    }
                    jogadores[qtJ].setId(Integer.parseInt(atributos[0]));
                    jogadores[qtJ].setNome(atributos[1]);
                    jogadores[qtJ].setAltura(Integer.parseInt(atributos[2]));
                    jogadores[qtJ].setPeso(Integer.parseInt(atributos[3]));
                    jogadores[qtJ].setUniversidade(atributos[4]);
                    jogadores[qtJ].setAnoNascimento(Integer.parseInt(atributos[5]));
                    jogadores[qtJ].setCidadeNascimento(atributos[6]);
                    jogadores[qtJ].setEstadoNascimento(atributos[7]);
                    qtJ++;// incremento da quantidade de jogadores registrados
                    sc.close();
                    
                }
                else qtJ++;
            }
            
        } catch (IOException e) {
            e.printStackTrace();
        }
        return jogadores;
    }
}
//Classe Celula
class Celula{
    public Celula prox;//APontador para uma outra Celula
    public Jogador elemento;
    public Celula(Jogador jogador){
        prox=null;
        this.elemento=jogador;
    }
    public Celula(){
        prox=null;
    }
} 
public class PilhaFlex{
    private Celula topo;
    public PilhaFlex(){
        topo=null;
    }
    //Método para realizar a insercao de um jogador na pilha
    public void inserir(Jogador jogador){
        Celula aux=new Celula(jogador);
        aux.prox=topo;
        topo=aux;
        aux=null;
    }
     //Método para calcular o numero de registros presentes na pilha 
    public int tamanho(){
        int c=0;
        Celula i=topo;
        while(i!=null){
            c++;
            i=i.prox;
        }
        return c;
    }
    public Jogador remover(){
        Jogador aux=topo.elemento;
        Celula x=topo;
        topo=x.prox;
        x=null;
        return aux;
    }
    //metodo para imprimir o array final
    public void imprime(Celula i,int j){
        if(i==null){
            return;
        }
        else{
            imprime(i.prox,j-1);
            System.out.println("["+j+"] ## "+i.elemento.getNome()+" ## "+i.elemento.getAltura()+" ## "+i.elemento.getPeso()+" ## "+i.elemento.getAnoNascimento()+" ## "+i.elemento.getUniversidade()+" ## "+i.elemento.getCidadeNascimento()+" ## "+i.elemento.getEstadoNascimento()+" ##");
        }
    }
    public static void main(String[] args) {
            
        Scanner scanner= new Scanner(System.in);
        String id="",comandos;
        int nOperacoes,qtRemovidos=0;
        PilhaFlex pilha=new PilhaFlex();
        Jogador jogadores[]=new Jogador[4000];
        Jogador removidos[]=new Jogador[100];
        jogadores=Jogador.ler();// Leitura dos jogadores no csv

        //Primeira parte de entradas
        do{
            id=scanner.next();
            if(!id.equals("FIM")){
                int identificador=Integer.parseInt(id);
                pilha.inserir(jogadores[identificador]);
            }            
        }while(!id.equals("FIM"));

        nOperacoes=scanner.nextInt();// numero de operacoes
        scanner.nextLine();
        for(int i=0;i<nOperacoes;i++){
            comandos=scanner.nextLine();// leitura de cada operacao

            //remocao
            if(comandos.charAt(0)=='R'){
                removidos[qtRemovidos]=pilha.remover();
                qtRemovidos++;
            }
            //insercao
            else if(comandos.charAt(0)=='I'){
                String[] palavras = comandos.split(" ");
                int identificador=Integer.parseInt(palavras[1]);
                pilha.inserir(jogadores[identificador]);
            }
        }
        //impressao dos removidos
        for(int i=0;i<qtRemovidos;i++){
            System.out.println("(R) "+removidos[i].getNome());
        }
        
        pilha.imprime(pilha.topo,pilha.tamanho()-1);
        scanner.close();
    }
    
}
